File: src/CalcViewModel/GraphingCalculator/GraphingCalculatorViewModel.cpp
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

#include "pch.h"
#include "GraphingCalculatorViewModel.h"

using namespace CalculatorApp::ViewModel;
using namespace Platform;
using namespace Platform::Collections;
using namespace Windows::Foundation;
using namespace Windows::Foundation::Collections;
using namespace Windows::UI::Xaml::Data;
using namespace GraphControl;

namespace CalculatorApp::ViewModel
{
    GraphingCalculatorViewModel::GraphingCalculatorViewModel()
        : m_IsDecimalEnabled{ true }
        , m_Equations{ ref new Vector<EquationViewModel ^>() }
        , m_Variables{ ref new Vector<VariableViewModel ^>() }
    {
    }

    void GraphingCalculatorViewModel::OnButtonPressed(Object ^ parameter)
    {
    }

    void GraphingCalculatorViewModel::UpdateVariables(IMap<String ^, Variable ^> ^ variables)
    {
        Variables->Clear();
        for (auto variablePair : variables)
        {
            auto variable = ref new VariableViewModel(variablePair->Key, variablePair->Value);
            variable->VariableUpdated += ref new EventHandler<VariableChangedEventArgs>([this, variable](Object ^ sender, VariableChangedEventArgs e) {
                VariableUpdated(variable, VariableChangedEventArgs{ e.variableName, e.newValue });
            });
            Variables->Append(variable);
        }
    }

    void GraphingCalculatorViewModel::SetSelectedEquation(EquationViewModel ^ equation)
    {
        SelectedEquation = equation;
    }
}

File: src/Calculator.ViewModels/GraphingCalculator/GraphingCalculatorViewModel.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

using System.Collections.Generic;
using System.Collections.ObjectModel;
using System.ComponentModel;
using CalculatorApp.ViewModel.Common;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;

namespace CalculatorApp.ViewModel
{
    public struct VariableChangedEventArgs
    {
        public string VariableName;
        public double NewValue;
    }

    [Windows.UI.Xaml.Data.Bindable]
    public sealed class GraphingCalculatorViewModel : ObservableObject
    {
        private bool _isDecimalEnabled;
        private ObservableCollection<EquationViewModel> _equations;
        private ObservableCollection<VariableViewModel> _variables;
        private EquationViewModel _selectedEquation;

        public event System.EventHandler<VariableChangedEventArgs> VariableUpdated;

        public GraphingCalculatorViewModel()
        {
            _equations = new ObservableCollection<EquationViewModel>();
            _variables = new ObservableCollection<VariableViewModel>();
            _isDecimalEnabled = true;
        }

        public bool IsDecimalEnabled
        {
            get => _isDecimalEnabled;
            private set => SetProperty(ref _isDecimalEnabled, value);
        }

        public ObservableCollection<EquationViewModel> Equations
        {
            get => _equations;
            private set => SetProperty(ref _equations, value ?? new ObservableCollection<EquationViewModel>());
        }

        public ObservableCollection<VariableViewModel> Variables
        {
            get => _variables;
            private set => SetProperty(ref _variables, value ?? new ObservableCollection<VariableViewModel>());
        }

        public EquationViewModel SelectedEquation
        {
            get => _selectedEquation;
            private set => SetProperty(ref _selectedEquation, value);
        }

        private RelayCommand<object> _buttonPressedCommand;
        public RelayCommand<object> ButtonPressedCommand => _buttonPressedCommand ?? (_buttonPressedCommand = new RelayCommand<object>(OnButtonPressed));
        public RelayCommand<object> ButtonPressed => ButtonPressedCommand;

        public void UpdateVariables(IEnumerable<KeyValuePair<string, GraphControl.Variable>> variables)
        {
            Variables.Clear();
            if (variables == null) return;

            foreach (var kvp in variables)
            {
                var vm = new VariableViewModel(kvp.Key, kvp.Value);
                vm.VariableUpdated += (sender, args) =>
                {
                    VariableUpdated?.Invoke(sender, args);
                };
                Variables.Add(vm);
            }
        }

        public void SetSelectedEquation(EquationViewModel equation)
        {
            SelectedEquation = equation;
        }

        private void OnButtonPressed(object parameter)
        {
            if (parameter is NumbersAndOperatorsEnum numOp)
            {
                // Handle graphing calculator button press
            }
        }

    }
}
